### PR TITLE
RichText: Ensure that hitting enter at the end of a paragraph creates an empty paragraph

### PR DESCRIPTION
**Pull request by youknowriad**
**Testing instructions**

 - Hit enter at the end of a paragraph
 - The side inserter should show up in the newly created paragraph

### CODE REVIEW

**Comment by aduth** [comment]
I've taken a few passes at refactoring this logic, which I've generalized to filtering out empty nodes when a split occurs, including empty text nodes and inline boundary placeholder nodes.

This resolves a few related bugs present in master:

1. Add a paragraph ending in a link
2. Add a paragraph following the first
3. From the start of the second paragraph, press backspace
4. Press enter

**Master:** An empty inline boundary fragment is included in the split paragraph

1. Add a paragraph ending in a code block
2. Move caret to the end of the code block
4. Press enter

**Master:** An empty inline boundary fragment is included in the split paragraph

**Comment by aduth** [comment]
Related: https://github.com/WordPress/gutenberg/issues/1246#issuecomment-327966045

**Comment by aduth** [comment]
Aside: `tinymce.DOM.isEmpty` behaves a bit strangely when it encounters this caret placeholder:

```js
tinymce.DOM.isEmpty( 'Is this empty? &#65279;' )
true
```

**Comment by iseulde** [comment]
Isn't the problem when a new RichText instance is created? I adjusted the check in the constructor and that seems to work: https://github.com/WordPress/gutenberg/pull/4956/files#diff-766381e792acf9d4c2505f3aeddcca7eR111

Where the problem is `[ '' ]` not being considered empty.

File: blocks/rich-text/index.js [hunk 1]
```diff
@@ -54,10 +55,52 @@ export function createTinyMCEElement( type, props, ...children ) {
...
+ * a placeholder caret character at the end.
+ *
+ * @see https://github.com/tinymce/tinymce/blob/master/src/plugins/link/main/ts/core/Utils.ts
+ *
+ * @param {Node} node Node to test.
+ *
+ * @return {boolean} Whether node is inline boundary.
+ */
+export function isEmptyInlineBoundary( node ) {
+	const text = node.nodeName === 'A' ? node.innerText : node.textContent;
```
Comment (iseulde) [inline blocks/rich-text/index.js]: This is not just on links anymore right? Could happen for bold etc. as well?

File: blocks/rich-text/index.js [hunk 2]
```diff
@@ -612,8 +655,7 @@ export class RichText extends Component {
 			const afterFragment = afterRange.extractContents();
 
 			const beforeElement = nodeListToReact( beforeFragment.childNodes, createTinyMCEElement );
-			const afterElement = isLinkBoundary( afterFragment ) ? [] : nodeListToReact( afterFragment.childNodes, createTinyMCEElement );
-
+			const afterElement = nodeListToReact( filterEmptyNodes( afterFragment.childNodes ), createTinyMCEElement );
```
Comment (iseulde) [inline blocks/rich-text/index.js]: Could the same not happen on `beforeElement`?

Comment (youknowriad) [reply blocks/rich-text/index.js]: The issue I'm seeing is that if the empty elements (`br`) are in the middle of the after content, they are also removed from the newly created block.

Comment (youknowriad) [reply blocks/rich-text/index.js]: In theory, this issue could happen in `beforeElement` when you trigger Enter at the beginning of a paragraph block but in my testings the block is considered Empty properly.

**Comment by aduth** [comment]
>Where the problem is `[ '' ]` not being considered empty.

Shouldn't we try to prevent this from ever occurring? That's what's being proposed in the latest changes here.

File: blocks/rich-text/index.js (hunk 1 above)
Comment (aduth) [reply blocks/rich-text/index.js]: Correct, in fact, this fixes the previous issue of only accounting for anchor. We test any kind of node, but from what I've tested and assuming from [the link plugin Utils.js](https://github.com/tinymce/tinymce/blob/a4add2940715dd27ef7be8ef42065a8292038787/src/plugins/link/main/ts/core/Utils.ts#L52-L53) is that it uses a different property to test (`innerText` for links, `textContent` for all others).

**Comment by aduth** [comment]
>Shouldn't we try to prevent this from ever occurring? That's what's being proposed in the latest changes here.

In fact, with these changes, plus this bit of logic:

https://github.com/WordPress/gutenberg/blob/4d835c0e873bb866b57561099c4b966a7ddfcb14/blocks/rich-text/index.js#L382

...it seems we shouldn't need to manage `empty` in state at all, and could just rely on:

```
const isEmpty = this.props.value.length === 0;
```

?

**Comment by aduth** [comment]
>The issue I'm seeing is that if the empty elements (`br`) are in the middle of the after content, they are also removed from the newly created block.

In 3476444, I updated `isEmptyNode` to check exclusively for the empty text node, not element nodes. I included a few unit additional unit tests for real-world scenarios where an empty text node might be present and gave it a few rounds of testing in the browser. These issues seem to be resolved now.

**Comment by iseulde** [comment]
@aduth Yes, ideally we would rely on the value for the `isEmpty` check, but there are many more cases where the content would be empty and not just an empty array. `[ <br>, '' ]` etc. If we can filter all those out through `getContent` and keep the content in sync, then great.

**Pull request merged**